src/util/retry.rs

```diff
@@ -0,0 +1,53 @@
+use std::collections::HashSet;
+use std::future::Future;
+use std::hash::Hash;
+use std::time::Duration;
+
+/// A function that performs and retries the given operation, acc. to a retry policy and a set of
+/// retriable errors.
+pub async fn retry<R, F, Fut, T, E>(
+	mut operation: F, retry_policy: &R, retriable_errors: Option<HashSet<E>>,
```

Comment:
Maybe we just pass a `Fn`? Feels a little more rust idiomatic, at least. Callers would likely just hardcode a few error checks without using a container.

Reply:
I disagree. That idiom used every in the rust standard library and is preferable to allocating a container, IMO.

Reply:
Come to think of it, in line with my other comments, which errors are retryable is really a policy, no? So instead of using a `Fn` this interface you could make a `RetryPolicy` for this, either taking a `Fn` (a la `Filter`), a set of errors, or individual errors.

```rust
ExponentialBackoffRetryPolicy::new(Duration::from_secs(2))
	.with_max_attempts(5)
	.retry_on_error(|e| matches!(e, VssError::InternalServerError(..)));
```

Reply:
Mentioned elsewhere, but wrapping all the retrier context in a struct and using that as an argument may be a good tradeoff. At least for the VSS example, needing a container and a custom `PartialEq` seems less ideal than simply using `matches!` in a closure.

Reply:
I don't really agree with this argument. If it's a library for general use, then it shouldn't impose requirements on what a policy should be. Users of the library can do so if they wish by decorating whatever policy their own users hand to them and state the constraints in the docs. Or they can hide the library behind their own interface and only expose what is applicable to their application.

Reply:
Unless it eventually succeeds... I think my point is that we should let the caller decide what their `RetryPolicy` should be. The `retry` method should be policy-agnostic. We can always add a warning to the documentation.

Reply:
You could have the `retry` interface take a `MaxAttemptsRetryPolicy` instead of a generic `RetryPolicy` if you want to enforce it at the interface level, I suppose.

Reply:
Hmm...

Option 3: Change `retry` to take `&MaxAttemptsRetryPolicy<R>` instead of `&R`.

Also, could add a `without_attempt_limit` provided function to `RetryPolicy` returning a `MaxAttemptsRetryPolicy`, changing the `max_attempts` field to an `Option` where it is set to `None`. That way, you can still enforce making a choice even if the user doesn't want a limit.